toolchain/check/cpp_thunk.cpp

```diff
@@ -0,0 +1,369 @@
+}
+
+// Returns true if a C++ thunk is required for the given type. A C++ thunk is
+// required for any type except for void, pointer types and signed 32-bit and
+// 64-bit integers.
+static auto IsThunkRequiredForType(Context& context, SemIR::TypeId type_id)
+    -> bool {
+  if (!type_id.has_value() || type_id == SemIR::ErrorInst::TypeId) {
+    return false;
+  }
+
+  type_id = context.types().GetUnqualifiedType(type_id);
+
+  switch (context.types().GetAsInst(type_id).kind()) {
+    case SemIR::PointerType::Kind: {
+      return false;
+    }
+
+    case SemIR::ClassType::Kind: {
+      if (!context.types().IsComplete(type_id)) {
+        // Signed integers of 32 or 64 bits should be completed when imported.
+        return true;
+      }
+
+      if (!context.types().IsSignedInt(type_id)) {
+        return true;
+      }
+
+      llvm::APInt bit_width =
+          context.ints().Get(context.types().GetIntTypeInfo(type_id).bit_width);
+      return bit_width != 32 && bit_width != 64;
+    }
+
+    default:
+      return true;
+  }
+}
+
+auto IsCppThunkRequired(Context& context, const SemIR::Function& function)
+    -> bool {
+  if (!function.clang_decl_id.has_value()) {
+    return false;
+  }
+
+  if (function.self_param_id.has_value()) {
+    // TODO: Support member methods.
+    return false;
+  }
+
+  SemIR::TypeId return_type_id =
+      function.GetDeclaredReturnType(context.sem_ir());
+  if (return_type_id.has_value()) {
+    // TODO: Support non-void return values.
+    return false;
+  }
+
+  bool thunk_required_for_param = false;
+  for (auto param_id :
+       context.inst_blocks().GetOrEmpty(function.call_params_id)) {
+    if (param_id == SemIR::ErrorInst::InstId) {
+      return false;
+    }
+    if (!thunk_required_for_param &&
+        IsThunkRequiredForType(
+            context,
+            context.insts().GetAs<SemIR::AnyParam>(param_id).type_id)) {
+      thunk_required_for_param = true;
+    }
+  }
+
+  return thunk_required_for_param;
+}
+
+// Returns whether the type is a pointer or a signed int of 32 or 64 bits.
+static auto IsScalarType(clang::ASTContext& ast_context, clang::QualType type)
+    -> bool {
+  if (type->isPointerType()) {
+    return true;
+  }
+
+  if (const auto* builtin_type = type->getAs<clang::BuiltinType>()) {
+    // TODO: Add support for other integer types.
+    if (builtin_type->getKind() == clang::BuiltinType::Int) {
+      uint64_t type_size = ast_context.getTypeSize(type);
+      return type_size == 32 || type_size == 64;
+    }
+  }
+
+  return false;
+}
+
+// Creates the thunk parameter types given the callee function. Also returns for
+// each type whether it is different from the matching callee function parameter
+// type.
+static auto BuildThunkParameterTypes(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl)
+    -> std::tuple<llvm::SmallVector<clang::QualType>,
+                  llvm::SmallVector<bool> > {
```

Comment:
Does clang-format insist on putting the space here?
```suggestion
                  llvm::SmallVector<bool>> {
```

Reply:
Done.